kms/capi/capi.go

```diff
@@ -0,0 +1,646 @@
+		}
+	case "RSA":
+		buf, err := nCryptExportKey(kh, BCRYPT_RSAPUBLIC_BLOB)
+		if err != nil {
+			return nil, fmt.Errorf("failed to export %v public key: %w", algGroup, err)
+		}
+		pub, err = unmarshalRSA(buf)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal %v public key: %w", algGroup, err)
+		}
+	default:
+		return nil, fmt.Errorf("unhandled algorithm group %v retrieved from key", algGroup)
+	}
+
+	return pub, nil
+}
+
+// New returns a new CAPIKMS.
+func New(ctx context.Context, opts apiv1.Options) (*CAPIKMS, error) {
+	var providerName string
+	containerName, err := randutil.UUIDv4()
+	providerName = "Microsoft Software Key Storage Provider"
+	pin := ""
+
+	if err != nil {
+		return nil, fmt.Errorf("could not generate UUIDv4: %w", err)
+	}
+
+	if opts.URI != "" {
+		u, err := uri.ParseWithScheme(Scheme, opts.URI)
+		if err != nil {
+			return nil, err
+		}
+
+		if v := u.Get(ProviderNameArg); v != "" {
+			providerName = v
+		}
+
+		pin = u.Pin()
+	}
+
+	// TODO: a provider is not necessary for certificate functions, should we move this to the key and signing functions?
+	ph, err := nCryptOpenStorage(providerName)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not open nCrypt provider: %w", err)
+	}
+
+	return &CAPIKMS{
+		providerName:   providerName,
+		providerHandle: ph,
+		containerName:  containerName,
+		pin:            pin,
+	}, nil
+}
+
+func init() {
+	apiv1.Register(apiv1.CAPIKMS, func(ctx context.Context, opts apiv1.Options) (apiv1.KeyManager, error) {
+		return New(ctx, opts)
+	})
+}
+
+func (k *CAPIKMS) Close() error {
+	if k.providerHandle != 0 {
+		return nCryptFreeObject(k.providerHandle)
+	}
+
+	return nil
+}
+
+// CreateSigner returns a new signer configured with the given signing key.
+func (k *CAPIKMS) CreateSigner(req *apiv1.CreateSignerRequest) (crypto.Signer, error) {
+	var u *uri.URI
+
+	u, err := uri.ParseWithScheme(Scheme, req.SigningKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "CreateSigner failed to parse URI")
+	}
+
+	if k.containerName = u.Get(ContainerNameArg); k.containerName == "" {
+		return nil, errors.Errorf("CreateSigner %v not specified", ContainerNameArg)
+	}
+
+	scPIN := u.Pin()
+
+	if scPIN == "" {
+		scPIN = k.pin
+	}
+
+	return newCAPISigner(k.providerHandle, k.containerName, scPIN)
+}
+
+// CreateKey generates a new key using Golang crypto and returns both public and
+// private key.
+func (k *CAPIKMS) CreateKey(req *apiv1.CreateKeyRequest) (*apiv1.CreateKeyResponse, error) {
+
+	if req.Name == "" {
+		return nil, errors.Errorf("capi signing request must have a name")
+	}
+
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	if k.containerName = u.Get(ContainerNameArg); k.containerName == "" {
+		// generate a uuid for the container name
+		k.containerName, err = randutil.UUIDv4()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate uuid: %w", err)
+		}
+	}
```

Comment:
This is weird, you should not update the `containerName` initialized in `New()`. I'm not sure but it looks like `k.containerName` might not be necessary and you just need to create one if the uri doesn't provide it.

If you need the same container name in `GetPublicKey` or `CreateSigner`, then the return `Name` should be an uri with all the necessary information to use it.

In general, the return should be something like this:

```
return &apiv1.CreateKeyResponse{
 	Name:      "capi:key=foo;foo=bar;...",
 	PublicKey: pub,
 	CreateSignerRequest: apiv1.CreateSignerRequest{
 		SigningKey: "capi:key=foo;foo=bar;...",
 	},
 }, nil
 ```